File: desktop-app-vue/src/main/plugins/plugin-loader.js
/**
 * PluginLoader - 插件加载器
 *
 * 职责：
 * - 支持多种插件来源（本地文件夹、NPM包、ZIP压缩包）
 * - 验证插件manifest
 * - 插件代码加载和安装
 */

const { logger } = require("../utils/logger.js");
const path = require("path");
const fs = require("fs");
const { app } = require("electron");
const { spawn } = require("child_process");

class PluginLoader {
  constructor() {
    // 插件安装目录
    this.pluginsDir = path.join(app.getPath("userData"), "plugins");
    // 临时目录
    this.tempDir = path.join(app.getPath("temp"), "chainlesschain-plugins");

    // 确保目录存在
    this.ensureDirectories();
  }

  /**
   * 确保必要的目录存在
   */
  ensureDirectories() {
    const dirs = [
      this.pluginsDir,
      path.join(this.pluginsDir, "official"),
      path.join(this.pluginsDir, "community"),
      path.join(this.pluginsDir, "custom"),
      this.tempDir,
    ];

    for (const dir of dirs) {
      if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir, { recursive: true });
      }
    }
  }

  /**
   * 解析插件来源
   * @param {string} source - 插件来源
   * @param {Object} options - 选项
   * @returns {Promise<string>} 插件路径
   */
  async resolve(source, options = {}) {
    logger.info(`[PluginLoader] 解析插件来源: ${source}`);

    // 1. 检查是否为本地路径
    if (fs.existsSync(source)) {
      const stat = fs.statSync(source);

      if (stat.isDirectory()) {
        logger.info(`[PluginLoader] 识别为本地目录: ${source}`);
        return source; // 开发模式：直接使用本地目录
      }

      if (stat.isFile() && source.endsWith(".zip")) {
        logger.info(`[PluginLoader] 识别为ZIP文件: ${source}`);
        // 解压ZIP到临时目录
        const extractPath = path.join(this.tempDir, `extract_${Date.now()}`);
        await this.extractZip(source, extractPath);
        return extractPath;
      }
    }

    // 2. 尝试作为NPM包处理
    if (this.isNpmPackage(source)) {
      logger.info(`[PluginLoader] 识别为NPM包: ${source}`);
      return await this.installFromNpm(source, options);
    }

    throw new Error(`无法解析插件来源: ${source}`);
  }

  /**
   * 判断是否为NPM包名
   * @param {string} source - 来源字符串
   * @returns {boolean}
   */
  isNpmPackage(source) {
    // NPM包名格式: package-name 或 @scope/package-name
    return /^(@[a-z0-9-~][a-z0-9-._~]*\/)?[a-z0-9-~][a-z0-9-._~]*$/.test(
      source,
    );
  }

  /**
   * 加载插件manifest
   * @param {string} pluginPath - 插件路径
   * @returns {Promise<Object>} manifest对象
   */
  async loadManifest(pluginPath) {
    const manifestPath = path.join(pluginPath, "plugin.json");

    if (!fs.existsSync(manifestPath)) {
      // 回退到 package.json
      const packagePath = path.join(pluginPath, "package.json");
      if (fs.existsSync(packagePath)) {
        return this.parsePackageJson(packagePath);
      }

      throw new Error("找不到 plugin.json 或 package.json");
    }

    const content = fs.readFileSync(manifestPath, "utf-8");
    const manifest = JSON.parse(content);

    // 验证必需字段
    this.validateManifest(manifest);

    return manifest;
  }

  /**
   * 从package.json解析插件manifest
   * @param {string} packagePath - package.json路径
   * @returns {Object} manifest对象
   */
  parsePackageJson(packagePath) {
    const content = fs.readFileSync(packagePath, "utf-8");
    const pkg = JSON.parse(content);

    // 检查是否有chainlesschain配置节
    if (!pkg.chainlesschain) {
      throw new Error("package.json中缺少chainlesschain配置节");
    }

    // 合并package.json和chainlesschain配置
    const manifest = {
      id: pkg.chainlesschain.id || pkg.name,
      name: pkg.name,
      version: pkg.version,
      author: pkg.author,
      description: pkg.description,
      homepage: pkg.homepage,
      license: pkg.license,
      main: pkg.main || "index.js",
      dependencies: pkg.dependencies || {},
      ...pkg.chainlesschain,
    };

    return manifest;
  }

  /**
   * 验证manifest
   * @param {Object} manifest - manifest对象
   * @throws {Error} 验证失败时抛出错误
   */
  validateManifest(manifest) {
    const required = ["id", "name", "version"];

    for (const field of required) {
      if (!manifest[field]) {
        throw new Error(`插件manifest缺少必需字段: ${field}`);
      }
    }

    // 验证版本号格式（简单的semver检查）
    if (!/^\d+\.\d+\.\d+/.test(manifest.version)) {
      throw new Error(`无效的版本号格式: ${manifest.version}，应为 x.y.z`);
    }

    // 验证插件ID格式
    if (!/^[a-z0-9.-]+$/.test(manifest.id)) {
      throw new Error(
        `无效的插件ID格式: ${manifest.id}，只能包含小写字母、数字、点和短横线`,
      );
    }

    logger.info(`[PluginLoader] Manifest验证通过: ${manifest.id}`);
  }

  /**
   * 安装插件到插件目录
   * @param {string} sourcePath - 源路径
   * @param {Object} manifest - manifest对象
   * @returns {Promise<string>} 安装后的路径
   */
  async install(sourcePath, manifest) {
    const category = manifest.category || "custom";
    const targetPath = path.join(this.pluginsDir, category, manifest.id);

    logger.info(`[PluginLoader] 安装插件: ${manifest.id} -> ${targetPath}`);

    // 如果目标已存在，先删除
    if (fs.existsSync(targetPath)) {
      logger.warn(`[PluginLoader] 目标路径已存在，将被覆盖: ${targetPath}`);
      await fs.promises.rm(targetPath, { recursive: true, force: true });
    }

    // 创建目标目录
    fs.mkdirSync(targetPath, { recursive: true });

    // 复制文件
    await this.copyDirectory(sourcePath, targetPath);

    // 安装NPM依赖（如果有）
    if (
      manifest.dependencies &&
      Object.keys(manifest.dependencies).length > 0
    ) {
      logger.info(`[PluginLoader] 安装NPM依赖...`);
      await this.installNpmDependencies(targetPath);
    }

    logger.info(`[PluginLoader] 插件安装成功: ${targetPath}`);

    return targetPath;
  }

  /**
   * 从NPM安装插件
   * @param {string} packageName - NPM包名
   * @param {Object} options - 选项
   * @returns {Promise<string>} 安装路径
   */
  async installFromNpm(packageName, options = {}) {
    const { version } = options;
    const versionSuffix = version ? `@${version}` : "";
    const fullPackage = `${packageName}${versionSuffix}`;

    const installPath = path.join(this.tempDir, `npm_${Date.now()}`);
    fs.mkdirSync(installPath, { recursive: true });

    logger.info(`[PluginLoader] 从NPM安装: ${fullPackage}`);

    // 执行 npm install
    await this.execCommand("npm", [
      "install",
      fullPackage,
      "--prefix",
      installPath,
    ]);

    // 返回安装后的路径
    const packagePath = path.join(installPath, "node_modules", packageName);

    if (!fs.existsSync(packagePath)) {
      throw new Error(`NPM安装失败，包不存在: ${packagePath}`);
    }

    return packagePath;
  }

  /**
   * 加载插件代码
   * @param {string} pluginPath - 插件路径
   * @returns {Promise<Object>} 插件代码信息
   */
  async loadCode(pluginPath) {
    const manifest = await this.loadManifest(pluginPath);
    const entryFile = manifest.main || "index.js";
    const entryPath = path.join(pluginPath, entryFile);

    if (!fs.existsSync(entryPath)) {
      throw new Error(`插件入口文件不存在: ${entryPath}`);
    }

    // 读取代码
    const code = fs.readFileSync(entryPath, "utf-8");

    return {
      code,
      entryPath,
      manifest,
    };
  }

  /**
   * 卸载插件
   * @param {string} pluginPath - 插件路径
   */
  async uninstall(pluginPath) {
    if (fs.existsSync(pluginPath)) {
      logger.info(`[PluginLoader] 卸载插件: ${pluginPath}`);
      await fs.promises.rm(pluginPath, { recursive: true, force: true });
      logger.info(`[PluginLoader] 插件已删除`);
    }
  }

  /**
   * 解压ZIP文件
   * @param {string} zipPath - ZIP文件路径
   * @param {string} extractPath - 解压目标路径
   */
  async extractZip(zipPath, extractPath) {
    // 使用 adm-zip 库（需要安装）
    try {
      const AdmZip = require("adm-zip");
      const zip = new AdmZip(zipPath);

      fs.mkdirSync(extractPath, { recursive: true });
      zip.extractAllTo(extractPath, true);

      logger.info(`[PluginLoader] ZIP解压成功: ${extractPath}`);
    } catch (error) {
      // 如果没有 adm-zip，使用系统命令
      if (process.platform === "win32") {
        await this.execCommand("powershell", [
          "-Command",
          `Expand-Archive -Path "${zipPath}" -DestinationPath "${extractPath}" -Force`,
        ]);
      } else {
        await this.execCommand("unzip", ["-q", zipPath, "-d", extractPath]);
      }
    }
  }

  /**
   * 复制目录
   * @param {string} src - 源目录
   * @param {string} dest - 目标目录
   */
  async copyDirectory(src, dest) {
    const entries = fs.readdirSync(src, { withFileTypes: true });

    for (const entry of entries) {
      const srcPath = path.join(src, entry.name);
      const destPath = path.join(dest, entry.name);

      // 跳过 node_modules（会重新安装）
      if (entry.name === "node_modules") {
        continue;
      }

      if (entry.isDirectory()) {
        fs.mkdirSync(destPath, { recursive: true });
        await this.copyDirectory(srcPath, destPath);
      } else {
        fs.copyFileSync(srcPath, destPath);
      }
    }
  }

  /**
   * 安装NPM依赖
   * @param {string} pluginPath - 插件路径
   */
  async installNpmDependencies(pluginPath) {
    return new Promise((resolve, reject) => {
      const npm = process.platform === "win32" ? "npm.cmd" : "npm";

      const child = spawn(npm, ["install", "--production"], {
        cwd: pluginPath,
        stdio: "pipe",
      });

      let output = "";

      child.stdout.on("data", (data) => {
        output += data.toString();
      });

      child.stderr.on("data", (data) => {
        output += data.toString();
      });

      child.on("close", (code) => {
        if (code === 0) {
          logger.info("[PluginLoader] NPM依赖安装成功");
          resolve();
        } else {
          logger.error("[PluginLoader] NPM依赖安装失败:", output);
          reject(new Error(`NPM依赖安装失败，退出代码: ${code}`));
        }
      });
    });
  }

  /**
   * 执行命令
   * @param {string} command - 命令
   * @param {Array} args - 参数
   * @returns {Promise<string>} 命令输出
   */
  async execCommand(command, args = []) {
    return new Promise((resolve, reject) => {
      const child = spawn(command, args, {
        stdio: "pipe",
        shell: true,
      });

      let stdout = "";
      let stderr = "";

      child.stdout.on("data", (data) => {
        stdout += data.toString();
      });

      child.stderr.on("data", (data) => {
        stderr += data.toString();
      });

      child.on("close", (code) => {
        if (code === 0) {
          resolve(stdout);
        } else {
          reject(new Error(`命令执行失败 (${code}): ${stderr || stdout}`));
        }
      });

      child.on("error", (error) => {
        reject(error);
      });
    });
  }
}

module.exports = PluginLoader;

File: ios-app/ChainlessChain/Features/Plugins/Services/PluginLoader.swift
import Foundation
import ZIPFoundation

// MARK: - PluginLoader
/// Plugin loading and installation
/// Ported from PC: plugins/plugin-loader.js
///
/// Features:
/// - Load plugins from various sources (local, ZIP, URL)
/// - Validate plugin manifests
/// - Install plugins to plugin directory
/// - Manage plugin dependencies
///
/// Version: v1.7.0
/// Date: 2026-02-10

// MARK: - Plugin Loader Errors

enum PluginLoaderError: Error, LocalizedError {
    case manifestNotFound
    case invalidManifest(String)
    case invalidVersion(String)
    case invalidPluginId(String)
    case entryFileNotFound(String)
    case unsupportedSource(String)
    case installationFailed(String)
    case unzipFailed(String)
    case downloadFailed(String)
    case dependencyMissing([String])
    case dependencyIncompatible(String)

    var errorDescription: String? {
        switch self {
        case .manifestNotFound:
            return "Cannot find plugin.json or package.json"
        case .invalidManifest(let reason):
            return "Invalid manifest: \(reason)"
        case .invalidVersion(let version):
            return "Invalid version format: \(version), should be x.y.z"
        case .invalidPluginId(let id):
            return "Invalid plugin ID: \(id), only lowercase letters, numbers, dots, and dashes allowed"
        case .entryFileNotFound(let path):
            return "Plugin entry file not found: \(path)"
        case .unsupportedSource(let source):
            return "Unsupported plugin source: \(source)"
        case .installationFailed(let reason):
            return "Plugin installation failed: \(reason)"
        case .unzipFailed(let reason):
            return "Failed to unzip plugin: \(reason)"
        case .downloadFailed(let reason):
            return "Failed to download plugin: \(reason)"
        case .dependencyMissing(let deps):
            return "Missing dependencies: \(deps.joined(separator: ", "))"
        case .dependencyIncompatible(let info):
            return "Incompatible dependency: \(info)"
        }
    }
}

// MARK: - Plugin Source Type

/// Types of plugin sources
enum PluginSourceType {
    case localDirectory(URL)
    case zipFile(URL)
    case remoteURL(URL)
    case bundled(String)
}

// MARK: - Plugin Loader

/// Loads and installs plugins from various sources
class PluginLoader {

    // MARK: - Properties

    private let logger = Logger.shared
    private let fileManager = FileManager.default

    /// Plugin installation directory
    let pluginsDirectory: URL

    /// Temporary directory for downloads/extracts
    let tempDirectory: URL

    // MARK: - Initialization

    init() {
        // Get app support directory
        let appSupport = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let appDirectory = appSupport.appendingPathComponent("ChainlessChain")

        // Set up directories
        self.pluginsDirectory = appDirectory.appendingPathComponent("plugins")
        self.tempDirectory = fileManager.temporaryDirectory.appendingPathComponent("chainlesschain-plugins")

        // Ensure directories exist
        ensureDirectories()
    }

    // MARK: - Directory Management

    /// Ensure required directories exist
    private func ensureDirectories() {
        let directories = [
            pluginsDirectory,
            pluginsDirectory.appendingPathComponent("official"),
            pluginsDirectory.appendingPathComponent("community"),
            pluginsDirectory.appendingPathComponent("custom"),
            tempDirectory
        ]

        for dir in directories {
            if !fileManager.fileExists(atPath: dir.path) {
                try? fileManager.createDirectory(at: dir, withIntermediateDirectories: true)
            }
        }
    }

    // MARK: - Source Resolution

    /// Resolve plugin source to a local path
    /// - Parameter source: Plugin source (path, URL, or bundle name)
    /// - Returns: Resolved local path
    func resolve(source: String) async throws -> URL {
        logger.info("[PluginLoader] Resolving source: \(source)")

        // Check if it's a local path
        let localURL = URL(fileURLWithPath: source)
        if fileManager.fileExists(atPath: source) {
            var isDirectory: ObjCBool = false
            fileManager.fileExists(atPath: source, isDirectory: &isDirectory)

            if isDirectory.boolValue {
                logger.info("[PluginLoader] Identified as local directory")
                return localURL
            }

            if source.hasSuffix(".zip") {
                logger.info("[PluginLoader] Identified as ZIP file")
                return try await extractZip(from: localURL)
            }
        }

        // Check if it's a remote URL
        if source.hasPrefix("https://") || source.hasPrefix("http://") {
            guard let url = URL(string: source) else {
                throw PluginLoaderError.unsupportedSource(source)
            }
            logger.info("[PluginLoader] Identified as remote URL")
            return try await downloadPlugin(from: url)
        }

        // Check if it's a bundled plugin
        if let bundleURL = Bundle.main.url(forResource: source, withExtension: nil, subdirectory: "Plugins") {
            logger.info("[PluginLoader] Identified as bundled plugin")
            return bundleURL
        }

        throw PluginLoaderError.unsupportedSource(source)
    }

    // MARK: - Manifest Loading

    /// Load and validate plugin manifest
    /// - Parameter pluginPath: Path to plugin directory
    /// - Returns: Parsed manifest
    func loadManifest(from pluginPath: URL) throws -> PluginManifest {
        let manifestPath = pluginPath.appendingPathComponent("plugin.json")
        let packagePath = pluginPath.appendingPathComponent("package.json")

        // Try plugin.json first
        if fileManager.fileExists(atPath: manifestPath.path) {
            let data = try Data(contentsOf: manifestPath)
            let manifest = try JSONDecoder().decode(PluginManifest.self, from: data)
            try validateManifest(manifest)
            return manifest
        }

        // Fall back to package.json
        if fileManager.fileExists(atPath: packagePath.path) {
            return try parsePackageJson(at: packagePath)
        }

        throw PluginLoaderError.manifestNotFound
    }

    /// Parse package.json as plugin manifest
    private func parsePackageJson(at path: URL) throws -> PluginManifest {
        let data = try Data(contentsOf: path)
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw PluginLoaderError.invalidManifest("Cannot parse package.json")
        }

        // Check for chainlesschain config section
        guard let chainlessConfig = json["chainlesschain"] as? [String: Any] else {
            throw PluginLoaderError.invalidManifest("Missing chainlesschain config in package.json")
        }

        // Build manifest from package.json + chainlesschain config
        let id = (chainlessConfig["id"] as? String) ?? (json["name"] as? String) ?? ""
        let name = (json["name"] as? String) ?? id
        let version = (json["version"] as? String) ?? "1.0.0"

        let manifest = PluginManifest(
            id: id,
            name: name,
            version: version,
            author: json["author"] as? String,
            description: json["description"] as? String,
            homepage: json["homepage"] as? String,
            license: json["license"] as? String,
            main: (json["main"] as? String) ?? "index.js",
            category: PluginCategory(rawValue: chainlessConfig["category"] as? String ?? "custom") ?? .custom,
            permissions: parsePermissions(chainlessConfig["permissions"]),
            dependencies: (json["dependencies"] as? [String: String]) ?? [:]
        )

        try validateManifest(manifest)
        return manifest
    }

    /// Parse permissions array
    private func parsePermissions(_ raw: Any?) -> [PluginPermission] {
        guard let array = raw as? [String] else { return [] }
        return array.compactMap { PluginPermission(rawValue: $0) }
    }

    /// Validate manifest required fields
    private func validateManifest(_ manifest: PluginManifest) throws {
        // Check required fields
        if manifest.id.isEmpty {
            throw PluginLoaderError.invalidManifest("Missing required field: id")
        }
        if manifest.name.isEmpty {
            throw PluginLoaderError.invalidManifest("Missing required field: name")
        }
        if manifest.version.isEmpty {
            throw PluginLoaderError.invalidManifest("Missing required field: version")
        }

        // Validate version format (x.y.z)
        let versionRegex = try! NSRegularExpression(pattern: #"^\d+\.\d+\.\d+"#)
        let versionRange = NSRange(manifest.version.startIndex..., in: manifest.version)
        if versionRegex.firstMatch(in: manifest.version, range: versionRange) == nil {
            throw PluginLoaderError.invalidVersion(manifest.version)
        }

        // Validate plugin ID format
        let idRegex = try! NSRegularExpression(pattern: #"^[a-z0-9.-]+$"#)
        let idRange = NSRange(manifest.id.startIndex..., in: manifest.id)
        if idRegex.firstMatch(in: manifest.id, range: idRange) == nil {
            throw PluginLoaderError.invalidPluginId(manifest.id)
        }

        logger.info("[PluginLoader] Manifest validation passed: \(manifest.id)")
    }

    // MARK: - Installation

    /// Install plugin to plugins directory
    /// - Parameters:
    ///   - sourcePath: Source plugin path
    ///   - manifest: Plugin manifest
    /// - Returns: Installed path
    func install(from sourcePath: URL, manifest: PluginManifest) throws -> URL {
        let category = manifest.category.rawValue
        let targetPath = pluginsDirectory
            .appendingPathComponent(category)
            .appendingPathComponent(manifest.id)

        logger.info("[PluginLoader] Installing plugin: \(manifest.id) -> \(targetPath.path)")

        // Remove existing if present
        if fileManager.fileExists(atPath: targetPath.path) {
            logger.warning("[PluginLoader] Target path exists, will be replaced")
            try fileManager.removeItem(at: targetPath)
        }

        // Create target directory
        try fileManager.createDirectory(at: targetPath, withIntermediateDirectories: true)

        // Copy files
        try copyPluginFiles(from: sourcePath, to: targetPath)

        logger.info("[PluginLoader] Plugin installed successfully: \(targetPath.path)")

        return targetPath
    }

    /// Copy plugin files excluding node_modules
    private func copyPluginFiles(from source: URL, to destination: URL) throws {
        let contents = try fileManager.contentsOfDirectory(at: source, includingPropertiesForKeys: nil)

        for item in contents {
            let itemName = item.lastPathComponent

            // Skip node_modules
            if itemName == "node_modules" {
                continue
            }

            let destItem = destination.appendingPathComponent(itemName)

            var isDirectory: ObjCBool = false
            fileManager.fileExists(atPath: item.path, isDirectory: &isDirectory)

            if isDirectory.boolValue {
                try fileManager.createDirectory(at: destItem, withIntermediateDirectories: true)
                try copyPluginFiles(from: item, to: destItem)
            } else {
                try fileManager.copyItem(at: item, to: destItem)
            }
        }
    }

    /// Uninstall plugin
    /// - Parameter pluginPath: Path to installed plugin
    func uninstall(at pluginPath: URL) throws {
        if fileManager.fileExists(atPath: pluginPath.path) {
            logger.info("[PluginLoader] Uninstalling plugin: \(pluginPath.path)")
            try fileManager.removeItem(at: pluginPath)
            logger.info("[PluginLoader] Plugin removed")
        }
    }

    // MARK: - ZIP Handling

    /// Extract ZIP file to temporary directory
    private func extractZip(from zipPath: URL) async throws -> URL {
        let extractPath = tempDirectory.appendingPathComponent("extract_\(Date().timeIntervalSince1970)")

        try fileManager.createDirectory(at: extractPath, withIntermediateDirectories: true)

        do {
            try fileManager.unzipItem(at: zipPath, to: extractPath)
            logger.info("[PluginLoader] ZIP extracted to: \(extractPath.path)")
            return extractPath
        } catch {
            throw PluginLoaderError.unzipFailed(error.localizedDescription)
        }
    }

    // MARK: - Download

    /// Download plugin from remote URL
    private func downloadPlugin(from url: URL) async throws -> URL {
        logger.info("[PluginLoader] Downloading plugin from: \(url)")

        let downloadPath = tempDirectory.appendingPathComponent("download_\(Date().timeIntervalSince1970).zip")

        do {
            let (data, response) = try await URLSession.shared.data(from: url)

            guard let httpResponse = response as? HTTPURLResponse,
                  httpResponse.statusCode == 200 else {
                throw PluginLoaderError.downloadFailed("HTTP error")
            }

            try data.write(to: downloadPath)
            logger.info("[PluginLoader] Downloaded to: \(downloadPath.path)")

            // Extract if it's a ZIP
            if url.pathExtension.lowercased() == "zip" {
                return try await extractZip(from: downloadPath)
            }

            return downloadPath
        } catch let error as PluginLoaderError {
            throw error
        } catch {
            throw PluginLoaderError.downloadFailed(error.localizedDescription)
        }
    }

    // MARK: - Code Loading

    /// Load plugin code information
    /// - Parameter pluginPath: Path to plugin
    /// - Returns: Code info including entry path and manifest
    func loadCode(from pluginPath: URL) throws -> PluginCodeInfo {
        let manifest = try loadManifest(from: pluginPath)
        let entryFile = manifest.main
        let entryPath = pluginPath.appendingPathComponent(entryFile)

        if !fileManager.fileExists(atPath: entryPath.path) {
            throw PluginLoaderError.entryFileNotFound(entryPath.path)
        }

        // Read code content
        let code = try String(contentsOf: entryPath, encoding: .utf8)

        return PluginCodeInfo(
            code: code,
            entryPath: entryPath,
            manifest: manifest
        )
    }

    // MARK: - Dependency Resolution

    /// Resolve plugin dependencies
    /// - Parameters:
    ///   - manifest: Plugin manifest
    ///   - installedPlugins: Currently installed plugins
    /// - Returns: Dependency resolution result
    func resolveDependencies(
        manifest: PluginManifest,
        installedPlugins: [String: PluginInfo]
    ) -> DependencyResolutionResult {
        guard !manifest.dependencies.isEmpty else {
            return DependencyResolutionResult(missing: [], incompatible: [], resolved: true)
        }

        logger.info("[PluginLoader] Resolving dependencies: \(manifest.dependencies.keys)")

        var missing: [PluginDependency] = []
        var incompatible: [(name: String, required: String, installed: String)] = []

        for (depName, versionConstraint) in manifest.dependencies {
            // Check if it's a plugin dependency
            if depName.hasPrefix("@chainlesschain/") || depName.hasPrefix("chainlesschain-") {
                if let installed = installedPlugins[depName] {
                    // Check version compatibility
                    if !isVersionCompatible(installed.manifest.version, with: versionConstraint) {
                        incompatible.append((
                            name: depName,
                            required: versionConstraint,
                            installed: installed.manifest.version
                        ))
                    }
                } else {
                    missing.append(PluginDependency(
                        pluginId: manifest.id,
                        dependencyId: depName,
                        type: .plugin,
                        version: versionConstraint
                    ))
                }
            }
            // System dependencies would be checked here
        }

        let resolved = missing.isEmpty && incompatible.isEmpty

        if !resolved {
            if !missing.isEmpty {
                logger.warning("[PluginLoader] Missing dependencies: \(missing.map { $0.dependencyId })")
            }
            if !incompatible.isEmpty {
                logger.warning("[PluginLoader] Incompatible dependencies: \(incompatible)")
            }
        }

        return DependencyResolutionResult(
            missing: missing,
            incompatible: incompatible.map { "\($0.name) (need \($0.required), have \($0.installed))" },
            resolved: resolved
        )
    }

    /// Simple version compatibility check
    private func isVersionCompatible(_ version: String, with constraint: String) -> Bool {
        // Simple implementation - exact match or prefix match
        if constraint.hasPrefix("^") {
            let major = constraint.dropFirst().components(separatedBy: ".").first ?? ""
            let versionMajor = version.components(separatedBy: ".").first ?? ""
            return major == versionMajor
        }
        if constraint.hasPrefix(">=") {
            let required = String(constraint.dropFirst(2))
            return compareVersions(version, required) >= 0
        }
        if constraint.hasPrefix(">") {
            let required = String(constraint.dropFirst())
            return compareVersions(version, required) > 0
        }
        return version.hasPrefix(constraint.replacingOccurrences(of: "x", with: ""))
    }

    /// Compare semantic versions
    private func compareVersions(_ v1: String, _ v2: String) -> Int {
        let parts1 = v1.components(separatedBy: ".").compactMap { Int($0) }
        let parts2 = v2.components(separatedBy: ".").compactMap { Int($0) }

        for i in 0..<max(parts1.count, parts2.count) {
            let p1 = i < parts1.count ? parts1[i] : 0
            let p2 = i < parts2.count ? parts2[i] : 0
            if p1 != p2 {
                return p1 - p2
            }
        }
        return 0
    }

    // MARK: - Cleanup

    /// Clean up temporary files
    func cleanupTemp() {
        do {
            let contents = try fileManager.contentsOfDirectory(at: tempDirectory, includingPropertiesForKeys: nil)
            for item in contents {
                try? fileManager.removeItem(at: item)
            }
            logger.info("[PluginLoader] Temp directory cleaned")
        } catch {
            logger.warning("[PluginLoader] Failed to clean temp: \(error)")
        }
    }
}

// MARK: - Supporting Types

/// Plugin code information
struct PluginCodeInfo {
    let code: String
    let entryPath: URL
    let manifest: PluginManifest
}

/// Dependency resolution result
struct DependencyResolutionResult {
    let missing: [PluginDependency]
    let incompatible: [String]
    let resolved: Bool
}
